pkg/apis/core/v1/helper/helpers.go

```diff
@@ -315,34 +316,45 @@ func NodeSelectorRequirementKeysExistInNodeSelectorTerms(reqs []v1.NodeSelectorR
 // MatchNodeSelectorTerms checks whether the node labels and fields match node selector terms in ORed;
 // nil or empty term matches no objects.
 func MatchNodeSelectorTerms(
-	nodeSelectorTerms []v1.NodeSelectorTerm,
-	nodeLabels labels.Set,
-	nodeFields fields.Set,
-) bool {
-	for _, req := range nodeSelectorTerms {
+	node *v1.Node,
+	nodeSelector *v1.NodeSelector,
+) (bool, error) {
+	if node == nil {
+		return false, nil
+	}
+	var errors []error
+	for _, req := range nodeSelector.NodeSelectorTerms {
 		// nil or empty term selects no objects
 		if len(req.MatchExpressions) == 0 && len(req.MatchFields) == 0 {
 			continue
 		}
 
 		if len(req.MatchExpressions) != 0 {
-			labelSelector, err := NodeSelectorRequirementsAsSelector(req.MatchExpressions)
-			if err != nil || !labelSelector.Matches(nodeLabels) {
+			labelSelector, err := nodeSelectorRequirementsAsSelector(req.MatchExpressions)
+			if err != nil {
+				errors = append(errors, err)
+				continue
+			}
+			if labelSelector == nil || !labelSelector.Matches(labels.Set(node.Labels)) {
```

Comment:
nit: do we need to check `labelSelector == nil`? (same for `fieldSelector == nil` at L349)

IIUC, `nodeSelectorRequirementsAsFieldSelector` always returns a nil selector along with an error.